design/vela-core/environment.md

```diff
@@ -1,294 +1,262 @@
-- Developers choose user Components and Environments to deploy Applications.
 
-When initializing environments, there are additional requirements:
+## Using KubeVela to initialize Environment
 
-- The ability to describe dependency relations between environments. For example, both *dev* and *prod* environments
-  would depend on a common environment that sets up operators both *dev* and *prod* would use.
+An environment is just a logical concept that groups a bundle of resources for multiple applications to share.
+To make an environment, we first need to initialize the shared resources.
 
-In the following we propose solutions to implement environment concept for application delivery.
+When initializing an environment, there is one additional requirement:
 
-## Initializer CRD
+- The ability to describe dependencies between environment modules. For example, both *dev* and *prod* environments
+  would depend on a common module that installs operators both *dev* and *prod* would use.
 
-We propose to add an Initializer CRD:
+To setup shared resources and define dependencies, we propose to use Application with two Workflow Steps
+to achieve the requirements. An example looks like below:
 
 ```yaml
-kind: Initializer
+apiVersion: core.oam.dev/v1beta1
+kind: Application
 metadata:
-  name: prod-env
+  name: setup-dev-env
 spec:
-  # appTemplate indicates the application template to render and deploy an system application.
-  appTemplate:
-    # We can use Application to deploy the resources that an environment needs.
-    # An environment setup can be done just like deploying applications in system level.
-    metadata:
-      ...
-    spec:
-      # system components
-      components:
-        ...
-      # system policies
-      policies:
-        ...
-
-  # dependsOn indicates the other initializers that this depends on.
-  # It will not apply its components until all dependencies exist.
-  dependsOn:
-    - ref:
-        apiVersion: core.oam.dev/v1beta1
-        kind: Initializer
-        name: common-operators
-```
+  # Using Application to deploy resources that makes up an environment
 
-By providing this initializer CRD, users can initialize environments by deploying system components --
-the abstractions is flexible enough to do work like provisioning cloud resources, handling db migration,
-installing system operators, etc.
-Additionally, initializer dependency enables separating common setup tasks like setting up namespaces,
-RBAC rules, accounts into reusable modules.
+  components:
+  - name: <env-component-name>
+    type: <env-component-type>
+    properties:
+      <env-component-props>
 
+  policies:
+  - name: <env-policy-name>
+    type: <env-policy-type>
+    properties:
+      <env-policy-props>
 
-## Initializer-Related Operators
+  workflow:
+  - name: wait-dependencies
+    # depends-on-app step will wait until the dependent app's status turns to Running phase
+    type: depends-on-app
```

Comment:
Is `workflow` in some application?

For this type, do all components of the application have to be ready first? Why not waiting for a component?

Reply:
Waiting for an app is good enough for initial use cases.
But definitely we can add more complex operations like waiting for a component.
We can wait until more use cases pops up and implement them later.